[WIP] 1.2 Test Split - Modules own tests

jaxwilko opened the pull request:
@LukeTowers told me to...

This PR splits out tests into their related modules, allowing for each module to be targeted using the `winter:test` command. Added to this is also proper namespacing and a bunch of minor bug fixes and tweaks.

jaxwilko committed:
Split tests by module and updated winter:test

jaxwilko committed:
Added js tests

jaxwilko committed:
Fixed js test pathing issues

jaxwilko committed:
Updated js test github action working dir

jaxwilko committed:
Added nl to improve readability

jaxwilko committed:
Code cleanup

jaxwilko committed:
Removed dev autoload as no longer required

jaxwilko committed:
Switched linting command for composer alias

jaxwilko committed:
Added snowboard tests to eslint ignore

jaxwilko committed:
Added backend module eslint ignore & comments

LukeTowers commented on modules/backend/tests/widgets/FormTest.php:
```diff
@@ -0,0 +1,243 @@
+        }
+    }
+
+
```
```suggestion
```

LukeTowers committed:
Update modules/backend/tests/widgets/FormTest.php

LukeTowers commented on modules/cms/tests/classes/AssetTest.php:
```diff
@@ -1,14 +1,17 @@
     {
         $theme = Theme::load('test');
-
+        
```
```suggestion

```

LukeTowers committed:
Update modules/cms/tests/classes/AssetTest.php

LukeTowers commented on modules/backend/tests/widgets/FormTest.php:
```diff
@@ -0,0 +1,242 @@
+namespace Backend\Tests\Widgets
+{
+    use System\Tests\Bootstrap\PluginTestCase;
+    use Illuminate\Database\Eloquent\Model;
```
```suggestion
    use Winter\Storm\Database\Model;
```

LukeTowers committed:
Update modules/backend/tests/widgets/FormTest.php

LukeTowers commented on modules/cms/tests/classes/ComponentManagerTest.php:
```diff
@@ -1,24 +1,28 @@
+use Cms\Classes\Layout;
+use Cms\Classes\Page;
+use Cms\Classes\Theme;
+use function base_path;
```
is this actually required or just a PHPStorm artifact?

jaxwilko replied:
Probably just phpstorm being helpful. Although technically I think we should be importing global functions but meh

LukeTowers replied:
Let's not import global functions. We don't do it anywhere else and it's not technically required right now.

jaxwilko replied:
Totally fine with that, i'll strip them out and push it up this evening :)

LukeTowers commented on modules/cms/tests/classes/PageTest.php:
```diff
@@ -11,7 +14,7 @@ public function testResolveMenuItem()
         $controller = new Controller;
 
-        $item = (object) [
+        $item = (object)[
```
```suggestion
        $item = (object) [
```

bennothommo commented:
Hey guys, 2 thoughts on this -

- Are we going to be forcing everyone to have to add `use System\Tests\Bootstrap\PluginTestCase` to their plugin test cases? I think we should probably still have that available as a global alias if possible.
- With the `tests` directories within the modules, we may have to look into using a `.gitattributes` file on each module with `export-ignore` definitions so that the tests aren't included on each install through Composer. (see https://github.com/composer/composer/issues/9153 and https://github.com/laravel/framework/blob/9.x/.gitattributes#L9-L23)

LukeTowers commented on modules/cms/tests/classes/RouterTest.php:
```diff
@@ -1,13 +1,18 @@
 use Cms\Classes\Router;
 use Cms\Classes\Theme;
+use ReflectionClass;
+use function count;
```
Another PHPStorm artifact?

jaxwilko replied:
As above

jaxwilko replied:
Although I prefer to `use ReflectionClass;` over `new \ReflectionClass();`. Your call though, not massively bothered.

LukeTowers replied:
I'm fine with the importing of the global classes, just not the functions.

LukeTowers commented on modules/cms/tests/classes/ThemeTest.php:
```diff
@@ -1,10 +1,17 @@
+use Config;
+use Event;
+use function base_path;
+use const PHP_OS_FAMILY;
```
PHPStorm?

LukeTowers committed:
Remove unnecessary docblock comments

LukeTowers committed:
Cleanup docblocks

LukeTowers commented on modules/system/tests/classes/ImageResizerTest.php:
```diff
@@ -82,7 +89,7 @@ public function testConfiguration()
 
         // Resize with an customised defaults
-        $this->markTestSkipped('@TODO: Something is going funky with serialized closures in PHP 8.0 / Laravel 9.x');
+//        $this->markTestSkipped('@TODO: Something is going funky with serialized closures in PHP 8.0 / Laravel 9.x');
```
```suggestion
```

LukeTowers commented on modules/system/tests/fixtures/themes/apitest/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*.htm
+testobjects/*.htm
```
```suggestion
testobjects/*.htm

```

LukeTowers committed:
Re-add README

LukeTowers committed:
Update modules/cms/tests/classes/PageTest.php

LukeTowers committed:
Update modules/system/tests/classes/ImageResizerTest.php

LukeTowers commented:
@bennothommo 

1. Yes, we are. Any existing projects will still have the original tests folder and they can keep using that if they want.
2. I want the tests folders for the modules distributed via Composer, that way the `winter:test` command can always be used and our testing suite can stay up to date with the rest of the platform. Right now if we make any changes to the testing suite or the base classes the only way users can get those changes is by manually downloading them to update their project tests folder.

bennothommo commented:
> @bennothommo
> 
>     1. Yes, we are. Any existing projects will still have the original tests folder and they can keep using that if they want.

As long as that doesn't create any conflicts, all good!

>     2. I want the tests folders for the modules distributed via Composer, that way the `winter:test` command can always be used and our testing suite can stay up to date with the rest of the platform. Right now if we make any changes to the testing suite or the base classes the only way users can get those changes is by manually downloading them to update their project tests folder.

I suppose it will be handy to not have to instruct people to download or updates their test folders anymore.

jaxwilko committed:
[WIP] Split Test - 1.2 - PluginManager lower case internal IDs (#583)

This modifies the plugin manager to internally treat plugin identifiers as lower case strings, only transforming them to their normalized versions when requested by methods like getPlugins() & getAllPlugins().

The idea behind this is that it provides a much simpler way to internally handle checking, especially for plugin replacement where casing could cause issues.

Replaces #576. Fixes #575.

jaxwilko committed:
Simplified getPlugins method key translation

jaxwilko committed:
Removed imports of global functions

jaxwilko commented:
Removed global func/const imports in d1ff0945d6959097f453fbb736d447a6efa2d966

jaxwilko committed:
Improved readability
